osquery/logger/plugins/kafka_producer.cpp

```diff
@@ -0,0 +1,308 @@
+#include <osquery/logger.h>
+#include <osquery/system.h>
+
+#include "osquery/config/parsers/kafka_topics.h"
+#include "osquery/logger/plugins/kafka_producer.h"
+
+namespace osquery {
+
+FLAG(string,
+     logger_kafka_brokers,
+     "localhost",
+     "Bootstrap broker(s) as a comma-separated list of host or host:port "
+     "(default port 9092)");
+
+FLAG(string, logger_kafka_topic, "", "Kafka topic to publish logs under");
+
+FLAG(string,
+     logger_kafka_acks,
+     "all",
+     "The number of acknowledgments the leader has to receive (0, 1, 'all')");
+
+/// Deleter for rd_kafka_t unique_ptr.
+void delKafkaHandle(rd_kafka_t* k) {
+  if (k != nullptr) {
+    rd_kafka_destroy(k);
+  }
+};
+
+/// Deleter for rd_kafka_topic_t unique_ptr.
+void delKafkaTopic(rd_kafka_topic_t* kt) {
+  if (kt != nullptr) {
+    rd_kafka_topic_destroy(kt);
+  }
+};
+
+REGISTER(KafkaProducerPlugin, "logger", "kafka_producer");
+
+std::once_flag KafkaProducerPlugin::shutdownFlag_;
+
+inline std::string getMsgName(const std::string& payload) {
+  const std::string fieldName = "\"name\"";
+  size_t ipos = payload.rfind(fieldName);
+  if (ipos == std::string::npos) {
+    // return base topic key
+    return "";
+  }
+
+  size_t first = payload.find('"', ipos + 6);
+  if (first == std::string::npos) {
+    return "";
+  }
+
+  size_t last = payload.find('"', first + 1);
+  if (last == std::string::npos) {
+    return "";
+  }
+
+  return payload.substr(first + 1, last - first - 1);
+}
+
+/**
+ * @brief callback for status of message delivery
+ *
+ * Logs an error message for failed deliveries; does nothing if successful.
+ * Callback is invoked by rd_kafka_poll.
+ */
+void onMsgDelivery(rd_kafka_t* rk,
+                   const rd_kafka_message_t* rkmessage,
+                   void* opaque) {
+  if (rkmessage->err != RD_KAFKA_RESP_ERR_NO_ERROR) {
+    LOG(ERROR) << "Kafka message delivery failed: "
+               << rd_kafka_err2str(rkmessage->err);
+  }
+}
+
+void KafkaProducerPlugin::flushMessages() {
+  WriteLock lock(producerMutex_);
+  rd_kafka_flush(producer_.get(), 3 * 1000);
+}
+
+void KafkaProducerPlugin::pollKafka() {
+  WriteLock lock(producerMutex_);
+  rd_kafka_poll(producer_.get(), 0 /*non-blocking*/);
+}
+
+void KafkaProducerPlugin::start() {
+  while (!interrupted() && running_.load()) {
+    pauseMilli(kKafkaPollDuration);
+    if (interrupted()) {
+      return;
+    }
+    pollKafka();
+  }
+}
+
+void KafkaProducerPlugin::stop() {
+  std::call_once(shutdownFlag_, [this]() {
+    if (running_.load()) {
+      running_.store(false);
+      flushMessages();
+    }
+  });
+}
+
+void KafkaProducerPlugin::init(const std::string& name,
+                               const std::vector<StatusLogLine>& log) {
+  // Get local hostname to use as client id and Kafka msg key.
+  std::string hostname(getHostname());
+
+  msgKey_ = hostname + "_" + name;
+
+  // Configure Kafka producer.
+  char errstr[512];
+  rd_kafka_conf_t* conf = rd_kafka_conf_new();
```

Comment:
The `make sanitize` should be in a good place now. If there's a test case that exercises this code path, it will tell you if there are leaks (90% sure it will).